server/subscribe.go

```diff
@@ -360,22 +367,140 @@ func (p *Plugin) listChannelSubscriptions() (string, error) {
+			var printChannelName = true
+			for _, subId := range subIds {
+				sub := subs.Channel.ById[subId]
+
+				channel, appErr := p.API.GetChannel(sub.ChannelId)
+				if appErr != nil {
+					return "", errors.New("Failed to get channel")
+				}
+
+				// only print channel name once for all subscriptions
+				if printChannelName {
+					rows = append(rows, fmt.Sprintf("* **~%s** (%d):", channel.Name, len(subIds)))
+					printChannelName = false
+				}
+
+				subName := "(No Name)"
+				if sub.Name != "" {
+					subName = sub.Name
+				}
+				rows = append(rows, fmt.Sprintf("  * %s - %s", sub.Filters.Projects.Elems()[0], subName))
+
+			}
+		}
+	}
+
+	return strings.Join(rows, "\n"), nil
+}
+
+func (p *Plugin) getSubscriptionsSorted() ([]TeamSubscriptions, error) {
+	subs, err := p.getSubscriptions()
+	if err != nil {
+		return nil, err
+	}
+
+	subsMap := make(map[string][][]string)
+	teamMap := make(map[string]string)
+
+	var nameIdArray []string
+
+	separator := "__subseparator__"
+	var DmSubsIds [][]string
+
+	// get teams from subscriptions
 	for channelID, subIDs := range subs.Channel.IdByChannelId {
+
+		// channel does not have any subIDs.
+		if len(subIDs) == 0 {
+			continue
+		}
+
 		channel, appErr := p.API.GetChannel(channelID)
 		if appErr != nil {
-			return "", errors.New("Failed to get channel")
+			return nil, errors.New("Failed to get channel")
 		}
 
-		rows = append(rows, fmt.Sprintf("~%s (%d):", channel.Name, len(subIDs)))
-
+		var channelSubIds []string
 		for subID := range subIDs {
-			sub := subs.Channel.ById[subID]
-			rows = append(rows, fmt.Sprintf("* %s - %s", sub.Filters.Projects.Elems()[0], sub.Name))
+			channelSubIds = append(channelSubIds, subID)
+		}
+
+		// for DMs and GMs, save to array and go to next team
+		if channel.TeamId == "" {
+			DmSubsIds = append(DmSubsIds, channelSubIds)
+			continue
+		}
+
+		// get need team.DisplayName for sorting
+		team, _ := p.API.GetTeam(channel.TeamId)
+
+		// teamMap used to determine if already have the team saved in nameIdArray
+		_, ok := teamMap[channel.TeamId]
+		if !ok {
+			teamNameId := team.DisplayName + separator + channel.TeamId
+			nameIdArray = append(nameIdArray, teamNameId)
 		}
+		teamMap[channel.TeamId] = team.DisplayName
+
+		// only save non-DM and non-GM subs to the map
+		subsMap[channel.TeamId] = append(subsMap[channel.TeamId], channelSubIds)
+
 	}
 
-	return strings.Join(rows, "\n"), nil
+	var teamSubs, dmSubs []TeamSubscriptions
+
+	// save all DM and GM channels under a generic teamName
+	if len(DmSubsIds) != 0 {
+		teamData := TeamSubscriptions{
+			TeamId:   "",
+			TeamName: "Group and Direct Messages",
+			SubIds:   DmSubsIds,
+		}
+		dmSubs = append(dmSubs, teamData)
+	}
+
+	// sort by combined team name/id values
+	sort.Strings(nameIdArray)
+
+	// save array of subs, sorted by team and grouped by channel
+	for _, teamId := range nameIdArray {
+
+		teamName := strings.Split(teamId, separator)[0]
+		teamId = strings.Split(teamId, separator)[1]
```

Comment:
Instead of using a string with a separator within it, which is sort of cryptic, can we use a struct that contains both values we want to extract?

Reply:
Yep.  After looking at it for a while, instead of creating any new `type`, I just save the `model.Team` types into an array.  Also, it isn't as easy to sort by struct fields in go, but found the included methodology, which was in the go documentation.  It leverages the `sort` library, but notice it requires satisfying the sort interface which means the `Len, Swap, and Less` methods must be present. 

Overall, it is more code but doesn't have the hacky string separator from before. Let me know what you think.

Reply:
Sounds good to me. Re-requesting @levb, to review these changes.